%YAML:1.0
# This file is based off the okvis configuration file for TUM VI dataset at
# https://vision.in.tum.de/_media/data/datasets/visual-inertial-dataset/config_okvis_50_20.yaml.tar.gz
# readout time = 29.4737 microseconds * 1024.
cameras:
     - {T_SC:
        [0.0, -1.0, 0.0, 0.0,
        -1.0, 0.0, 0.0,  0.0,
         0.0, 0.0, -1.0, 0.0,
         0.0, 0.0, 0.0,  1.0],
        image_dimension: [1280, 1024],
        distortion_coefficients: [0.0, 0.0, 0.0, 0.0],
        distortion_type: equidistant,
        focal_length: [740.0, 740.0],
        principal_point: [640.0, 512.0],
        down_scale: 2,
        projection_opt_mode: FX_CXY,
        extrinsic_opt_mode: P_CB,
        image_delay: 0.0,
        image_readout_time: 0.02}
     - {T_SC:
        [0.0, -1.0,  0.0, 0.0,
        -1.0,  0.0,  0.0, 0.0,
         0.0,  0.0, -1.0, 0.0,
         0.0,  0.0,  0.0, 1.0],
        image_dimension: [1280, 1024],
        distortion_coefficients: [0.0, 0.0, 0.0, 0.0],
        distortion_type: equidistant,
        focal_length: [740.0, 740.0],
        principal_point: [640.0, 512.0],
        down_scale: 2,
        projection_opt_mode: FX_CXY,
        extrinsic_opt_mode: P_C0C_Q_C0C,
        image_delay: 0.0,
        image_readout_time: 0.02}

monocular_input: false

camera_params:
    camera_rate: 20 # just to manage the expectations of when there should be frames arriving
    sigma_absolute_translation: 0.03 # The standard deviation of the camera extrinsics translation, e.g. 1.0e-10 for online-calib [m].
    sigma_absolute_orientation: 0.02 # The standard deviation of the camera extrinsics orientation, e.g. 1.0e-3 for online-calib [rad].
    sigma_c_relative_translation: 1.0e-6 # The std. dev. of the cam. extr. transl. change between frames, e.g. 1.0e-6 for adaptive online calib (not less for numerics) [m].
    sigma_c_relative_orientation: 1.0e-6 # The std. dev. of the cam. extr. orient. change between frames, e.g. 1.0e-6 for adaptive online calib (not less for numerics) [rad].
    timestamp_tolerance: 0.005 # [s] stereo frame out-of-sync tolerance
    sigma_focal_length: 5.0
    sigma_principal_point: 5.0
    sigma_distortion: [0.1, 0.01, 0.01, 0.01, 0.0]
    sigma_td: 0.1
    sigma_tr: 0.05

imu_params:
    a_max: 176.0 # acceleration saturation [m/s^2]
    g_max: 7.8 # gyro saturation [rad/s]
    sigma_g_c: 0.004  # gyro noise density [rad/s/sqrt(Hz)]
    sigma_a_c: 0.07  # accelerometer noise density [m/s^2/sqrt(Hz)]
    sigma_bg: 0.03 # gyro bias prior [rad/s]
    sigma_ba: 0.1 # accelerometer bias prior [m/s^2]
    sigma_gw_c: 0.000044 # gyro drift noise density [rad/s^s/sqrt(Hz)]
    sigma_aw_c: 0.00172 # accelerometer drift noise density [m/s^2/sqrt(Hz)]
    tau: 3600.0 # reversion time constant, currently not in use [s]
    # Local gravity at 85748 Garching, Germany (latitude: 48.262768, elevation: 481m), according to https://www.sensorsone.com/local-gravity-calculator/
    g: 9.80766 # Earth's acceleration due to gravity [m/s^2]
    a0: [ 0.0, 0.0, 0.0 ] # Accelerometer bias [m/s^2]
    imu_rate: 200
    # tranform Body-Sensor (IMU)
    T_BS:
        [1.0000, 0.0000, 0.0000, 0.0000,
         0.0000, 1.0000, 0.0000, 0.0000,
         0.0000, 0.0000, 1.0000, 0.0000,
         0.0000, 0.0000, 0.0000, 1.0000]
    sigma_TGElement: 5.0e-3
    sigma_TSElement: 2.0e-3
    sigma_TAElement: 5.0e-3
    model_type: BG_BA_TG_TS_TA

initial_state:
    std_v_WS: [0.1, 0.1, 0.1]
    std_q_WS: [0.2, 0.2, 0.05]
    std_p_WS: [0.1, 0.1, 0.1]

# Estimator parameters
numKeyframes: 5 # number of keyframes in optimisation window
numImuFrames: 3 # number of frames linked by most recent nonlinear IMU error terms
optimization:
    triangulationMaxDepth: 1000.0
    algorithm: "HybridFilter"
    useEpipolarConstraint: false
    cameraObservationModelId: 0
    computeOkvisNees: 0
    useMahalanobisGating: 1
    maxProjectionErrorTol: 7

frontend:
    stereoMatchWithEpipolarCheck: true
    epipolarDistanceThreshold: 2.5
    featureTrackingMethod: 0

point_landmark:
    landmarkModelId: 1
    minTrackLength: 3
    minTrackLengthForSlam: 7
    maxHibernationFrames: 3
    maxInStateLandmarks: 50
    maxMarginalizedLandmarks: 50

pose_graph:
    maxOdometryConstraintForAKeyframe: 2

# ceres optimization options
ceres_options:
    minIterations: 3   # minimum number of iterations always performed
    maxIterations: 10  # never do more than these, even if not converged
    timeLimit: -1      # [s] negative values will set the an unlimited time limit

# detection
detection_options:
    threshold: 40.0      # detection threshold. By default the uniformity radius in pixels
    octaves: 0           # number of octaves for detection. 0 means single-scale at highest resolution
    maxNoKeypoints: 400  # restrict to a maximum of this many keypoints per image (strongest ones)

# delay of images [s]:
imageDelay: 0.0  # in case you are using a custom setup, you will have to calibrate this. 0 for the VISensor.

# display debug images?
displayImages: false  # displays debug video and keyframe matches. May be slow.

# use direct driver
useDriver: false

# some options for how and what to publish -- optional in ROS-free version
publishing_options:
    publish_rate: 200                  # rate at which odometry updates are published only works properly if imu_rate/publish_rate is an integer!!
    publishLandmarks: false            # select, if you want to publish landmarks at all
    landmarkQualityThreshold: 1.0e-2   # landmark with lower quality will not be published
    maximumLandmarkQuality: 0.05       # landmark with higher quality will be published with the maximum colour intensity
    maxPathLength: 20                  # maximum length of the published path
    publishImuPropagatedState: false    # Should the state that is propagated with IMU messages be published? Or just the optimized ones?
    # provide custom World frame Wc
    T_Wc_W:
        [1.0000, 0.0000, 0.0000, 0.0000,
         0.0000, 1.0000, 0.0000, 0.0000,
         0.0000, 0.0000, 1.0000, 0.0000,
         0.0000, 0.0000, 0.0000, 1.0000]
    trackedBodyFrame: B                # B or S, the frame of reference that will be expressed relative to the selected worldFrame
    velocitiesFrame: Wc                # Wc, B or S,  the frames in which the velocities of the selected trackedBodyFrame will be expressed in

# gtsam smoother options
backend_modality: 0 # 0 structureless factors, 1 projection factors.
smartNoiseSigma: 1.5
rankTolerance: 1.0
landmarkDistanceThreshold: 20.0

outlierRejection: 8.0
retriangulationThreshold: 1.0e-3
addBetweenStereoFactors: 1

betweenRotationPrecision: 0.0
betweenTranslationPrecision: 100

relinearizeThreshold: 1.0e-2
relinearizeSkip: 1.0

numOptimize: 1
wildfire_threshold: 0.001
useDogLeg: 0

initialLambda: 0.0    # same as GN
lowerBoundLambda: 0.0 # same as GN
upperBoundLambda: 0.0 # same as GN

zeroVelocitySigma: 1.0e-3
noMotionPositionSigma: 1.0e-3
noMotionRotationSigma: 1.0e-4
constantVelSigma: 1.0e-2
